### DISCUSSION TITLE
FoldingRangeProvider reset all markers for a file

### DISCUSSION THREAD

**Discussion by nipatriknilsson** [microsoft/vscode-discussions#1217] [Extension Development QnA] [answered]
I'm using FoldingRangeProvider to fold a document. How can I reset all markers in a file, effectively erasing all markers and start over again? So all providers are called from scratch again.

It doesn't help to just return an empty array.

Sometimes there are some left-over-markers since last time the document was saved. I want to get rid of them.

**Comment by tamuratak** [comment]
You have to fire the `FoldingRangeProvider.onDidChangeFoldingRanges` event to tell VS Code that the folding ranges have changed.  It would be like

```ts
export class FoldingProvider implements vscode.FoldingRangeProvider {
    private emitter = new vscode.EventEmitter<void>()
    readonly onDidChangeFoldingRanges = this.emitter.event
```

and, you call `foldingProvider.emitter.fire()` when needed.

**Reply by nipatriknilsson** [reply]
Thank you for your reply!

Unfortunately it doesn't work. I have tried your code and this as well, which repeatedly fires...
```
_onDidChangeFoldingRangeEmitter = new vscode.EventEmitter<void>();
onDidChangeFoldingRanges = this._onDidChangeFoldingRangeEmitter.event;

constructor() {
	setInterval(() => {
		this._onDidChangeFoldingRangeEmitter.fire();
	}, 1000);
}
```

It seems depend on the saved state of the file.

While experimenting the following sequence works sometimes:

1. Save the file
2. Make the provider return []
3. Induce a change
4. Save

Sometimes this works better:

1. Save the file
2. Make the provider return []
3. Set the file type to ignore: vscode.languages.setTextDocumentLanguage(editor.document, 'ignore');
4. Induce a change
5. Save
6. Set the file type to the original.
7. Save

Nothing really reliable though.

**Comment by tamuratak** [comment, marked as answer]
It sounds like a bug. You had better report to https://github.com/Microsoft/vscode. For the report, it would be better to make clear whether `Save the file` means editing and saving the file with VS Code or not, e.g., a dev tool makes the change.
